docs/plugin/interfaces/guidelines.md

```diff
@@ -0,0 +1,383 @@
+
+            if( browser != nullptr ) {
+                ...
+                browser->Release();
+            }
+        ...
+```
+if the component does not expose the requested interface a nullptr will be returned and logic can be created around the availability of the interface.
+
+This feature of COM-RPC can be used in multiple ways:
+
+* When designing an interface make sure to only give it one distinct responsibility per interface. This makes it easier to understand and maintain the interface but now also gives a component the opportunity to only implement one responsibility and not the other if it does not make sense. The client can, using QueryInterface, check if also the extended functionality is available. For example for a Player interface where you could also change the Volume one would  create a separate IPlayer interface as well as an IVolume, instead of only having an IPlayer interface that would provide both. This will also make reuse of an interface easier as it can be used for any component implementing that responsibility , no reason to say it does not fit the component as only some part of the interface applies to what it is trying to achieve.
+
+* When an interface needs to be extended instead of adding new methods to the existing one, one could add the extended functionality to a new interface. Now the client can also work with components that do not yet implement the new extended functionality (as it can find out with QueryInterface if it is available) while on the other hand you are not forcing an update of all the components that do implement the initial interface, you only update the one(s) for which the extended new functionality makes sense.
+
+Design interfaces to be asynchronous when needed {#AsynchronousInterfaces}
+------------------------------------------------
+
+Sometimes when designing an interface it can be expected that the implementation of a certain action can be quite expensive and take some time. 
+
+In such a case (we usually use an expected duration of a few hundred milliseconds as the limit) or when the duration of an action is non deterministic and worse case would take longer than a few hundred milliseconds, it is advised to make the interface asynchronous. The interface will then only pass the data needed to start the action to the component and add a callback to the interface to make it possible for the implementation to return the result asynchronously. The component will handle the action in the background (do not use threads for this but use the Thunder workerpool) and call the provided callback when the action is completed to signal this to the initiator of the action. Not only will this prevent the call to fail because of JSON-RPC or COM-RPC timeouts but it will also make sure the framework is not blocked internally and also the client can continue without being blocked waiting for the result (or needed to reserve a thread for this).
```

Comment:
I would word the limit on method execution times more strongly.
e.g. Any call could take more than 300ms must use an asynchronous pattern.

Probably worth making developers aware of the JSON-RPC and COM-RPC timeout values.